[Windows] Fix chroma upsampling for AMD dxva processor and 10 bit output of SDR sources

CrystalP opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->

The quality of chroma upsampling is problematic with the AMD dxva processor for certain combinations.

8/10 bit YCbCr SDR to 10 bit RGB: OK luma, low quality chroma upsampling (point)
8/10 bit YCbCr SDR to 8 bit RGB: OK
10 bit YCbCr HDR/HLG to 8/10 SDR or HDR: OK
(using the latest driver 23.11.1 / 31.0.21905.1001)

Solution: force 8 bit output of the processor for SDR source.
Done with a change of the rules selecting the format of the intermediate target.

No such problem found with Intel and NVIDIA.

For future reference (float/linear not used yet in Kodi)
8 bit YCbCr SDR to float/linear RGB: bad, point luma and chroma
10 bit YCbCr SDR or HDR to float/linear RGB: OK

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

Triggered by investigation of chroma issues with pixel shaders and software render methods, prompted by #21850.

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Windows 8 and 10, x64 and UWP versions. Intel 4th, 8th, 13th gen, AMD, NVIDIA.
1080p, 720p, 640x380 videos with sharp red characters or logos. codec doesn't matter.
8 bit and 10 bit sources.

[8 bit sample](https://github.com/xbmc/xbmc/assets/489377/dd90f6b3-baa5-4d62-a595-b8988dcbbe72) [10 bit sample](https://github.com/xbmc/xbmc/assets/489377/1aa5eed7-566f-47b7-b1a7-00e18a8819d7)
Zoom 2x to see the impact on chroma very easily.

Not tested on xbox, equipped with an AMD gpu. Not sure that it has the issue in the first place.

## What is the effect on users?
<!--- Summarize the effect of this change on Kodi end-users. -->
<!--- If the PR does not have a noticeable impact (e.g., if it only changes documentation), -->
<!--- just leave it empty. Put in more detail the bigger the impact is. -->
<!--- This section may be used for automatic creation of release notes. -->
Better picture quality with the dxva render method on AMD for SDR material (used by default with hardware decoding)

## Screenshots (if appropriate):
Using the example of the issue, current master:
![image](https://github.com/xbmc/xbmc/assets/489377/ed519e42-3ce3-40e1-a456-f5270eae3a20)

with PR:
![image](https://github.com/xbmc/xbmc/assets/489377/e78c5bd5-de1b-4b99-9c00-abef4d71cb5a)

## Types of change
<!--- What type of change does your code introduce? Put an `x` with no space in all the boxes that apply like this: [X] -->
- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` with no space in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

CrystalP commented:
Added short sample in description for reproduction.

thexai commented:
Tested samples on Xbox (current master) and definitively is NOT affected by this issue in DXVA whereas that is very visible in Pixel Shaders path.

Also original user that reported this issue always has said not happens with DXVA path (DXVA2 + DXVA), only (ffmpeg + Pixel Shaders) and he has AMD hardware too:

> This happens all of my PC with different GPUs: Intel HD4400, AMD Vega8, AMD 5700xt and Nvidia GTX1660Super. I don't think all of them are out of date.

Then seems only very specific AMD setups has this issue and not seems good option disable for all AMD. Maybe advancesettings? But not seems necessary to me because users already has GUI settings "Use 10 bit for SDR" and "Use high precision processing" to able disable this on affected setups or due user personal preference (priority for 8-bit croma upsampling or better color precision/less banding).

Maybe this is only regression in very recent AMD drivers and in older drivers (and future drivers) this not happens.

thexai commented on xbmc/cores/VideoPlayer/VideoRenderers/windows/RendererDXVA.cpp:
```diff
@@ -484,15 +484,26 @@ bool CRendererDXVA::CRenderBufferImpl::UploadToTexture()
+  // As a workaround, fall back to 8 bit RGB output.
+
+  const bool force8bit = DX::DeviceResources::Get()->GetAdapterDesc().VendorId == PCIV_AMD &&
+                         !(conversionArgs.m_colorPrimaries == AVCOL_PRI_BT2020 &&
```
This excludes HDR sources but what happens with 10 bits SDR? (x264 / H.264 10-bit). 

Should accurate chroma or preserve 10-bit precision take priority? is not clear to me.

CrystalP replied:
On purpose, the only thing that triggers the bug is SDR converted to 10 bit RGB. HDR to 10 bit is fine.

For SDR, I'm not aware of real content with 10 actual bits of precision, they're at least 99% artificial 10bit for improved encoding efficiency purposes. The chroma artifacts are much more noticeable to me than a difference of value of 1 on a pixel sometimes.

CrystalP commented:
> Tested samples on Xbox (current master) and definitively is NOT affected by this issue in DXVA whereas that is very visible in Pixel Shaders path.
> 
> Also original user that reported this issue always has said not happens with DXVA path (DXVA2 + DXVA), only (ffmpeg + Pixel Shaders) and he has AMD hardware too:

I also have an AMD Vega 8 igpu (5600G), the same as the user, and the problem is definitely there, and I know why he didn't see it.

My PR "High precision processing" made 10 bit intermediate target the default. Before that, it was following the backbuffer bit-ness.
At the time of the report, AMD already had "Never" as the default value for "Use 10 bit for SDR", and the intermediate target would have been 8 bit. And with 8 bit IT, there is no problem, per my tests.

We don't have his log for proof, they disappear after a few months, but that's very likely.

When you tested the X-box did you make sure that the intermediate target was 10 bit (easily seen with alt-O)? That's when the problem happens.

thexai commented:
> And with 8 bit IT, there is no problem, per my tests.

Then basically you says that AMD has issues with 10-bit intermediate target... and users ALREADY have a GUI setting to disable 10 bit for SDR.

CrystalP commented:
Yes, problem with 10-bit intermediate target, with SDR source.

Yes technically there are settings to avoid the situation, but I wouldn't say that they're obvious and default settings should not cause problems (within reason...).
For Intel and NVIDIA, the default values of "Use 10 bit for SDR" (auto = on) and "Use high precision processing" (on) give good results.

The PR overrides the "Use high precision processing" in the known bad situation of AMD.

thexai commented:
But we don't know yet if all AMD systems are affected or just a small part. In fact, we already know that at least Xbox is not affected and this PR would prevent the use of 10-bit IT.

It's better as is now that the users can choose than to force it off by overriding the GUI setting silently (in all AMD even is not affected).

CrystalP commented:
True. It's not a hw problem but is the driver (AMD thought some conversions are not commonly used and didn't optimize), but couldn't say why xbox isn't affected. And there are different xbox... Maybe it's a different driver branch for xbox vs desktop Windows, or is related to gpu generation.
Would need some additional feedback, but it took 5 years for someone to notice the issue in pixel shaders and say something...

CrystalP committed:
[Windows] Functions to parse driver version string and compare them

CrystalP committed:
[Windows][DXVA] workaround low quality chroma  of AMD GCN gpu

Affects sources that don't use BT2020 primaries and 10 bit output.
RDNA architecture drivers don't list those conversions as supported, this change does the same for the GCN architecture.

CrystalP commented:
Thanks to testing by OP in original issue, the problem was narrowed down to the GCN architecture.
Changed the code to test those specifically (driver is < 31.0.22000.0) and not exclude the xbox.
Moved the code from dxva renderer to the enumeration of supported conversions, where it makes more sense.

thexai commented on xbmc/platform/win32/WIN32Util.cpp:
```diff
@@ -1629,9 +1629,19 @@ VideoDriverInfo CWIN32Util::GetVideoDriverInfo(const UINT vendorId, const std::w
+      // copy of the version, the original is destroyed in the process
+      std::string v{info.version};
+
+      char* p = strtok(v.data(), ".");
```
```suggestion
      char* p = std::strtok(v.data(), ".");
```

thexai commented on xbmc/platform/win32/WIN32Util.cpp:
```diff
@@ -1629,9 +1629,19 @@ VideoDriverInfo CWIN32Util::GetVideoDriverInfo(const UINT vendorId, const std::w
+      for (int idx = 0; p && idx < 4; ++idx)
+      {
+        info.part[idx] = std::stoi(p);
+        p = strtok(NULL, ".");
```
```suggestion
        p = std::strtok(nullptr, ".");
```

thexai commented on xbmc/rendering/dx/DeviceResources.cpp:
```diff
@@ -1484,3 +1484,20 @@ bool DX::DeviceResources::SetMultithreadProtected(bool enabled) const
+  if (GetAdapterDesc().VendorId != PCIV_AMD)
+    return false;
+
+  VideoDriverInfo driver = DX::DeviceResources::Get()->GetVideoDriverVersion();
```
```suggestion
  const VideoDriverInfo driver = GetVideoDriverVersion();
```

thexai replied:
`GetVideoDriverVersion()`is a member method of `DeviceResources`. 

Why you need `DX::DeviceResources::Get()->` to use it internally?

CrystalP replied:
It was initially in a different class. Replaced with a direct call to `CWIN32Util::GetVideoDriverInfo()` to avoid double call to GetAdapter and not necessary logging.

thexai commented on xbmc/rendering/dx/DeviceResources.cpp:
```diff
@@ -1484,3 +1484,20 @@ bool DX::DeviceResources::SetMultithreadProtected(bool enabled) const
+  VideoDriverInfo driver = DX::DeviceResources::Get()->GetVideoDriverVersion();
+
+  if (driver.part[0] > 31 || (driver.part[0] == 31 && driver.part[1] > 0) ||
+      (driver.part[0] == 31 && driver.part[1] > 0 && driver.part[2] >= 22000))
```
`driver.part[0] == 31 && driver.part[1] > 0` seems redundant with previous condition

You mean `(driver.part[0] == 31 && driver.part[1] == 0 && driver.part[2] >= 22000)` ?

CrystalP replied:
I don't think so, eventually there will be a new WDDM with 32.xx.xx.xx versions, I don't know if there will ever be 31.1, ...
I'd like to compact the expression but not sure how in a safe way.

That issue was side-stepped in IsWindowsVersionAtLeast() by assigning our own identifiers, which is not doable here.
Maybe a dedicated function to test IsVersionAtLeast() would make this more presentable.

thexai replied:
(A & B) | (A & B & C) is same as (A & B) but is not what I'm saying. For me seems a typo and current code:

```c++
if (driver.part[0] > 31 || (driver.part[0] == 31 && driver.part[1] > 0) ||
      (driver.part[0] == 31 && driver.part[1] > 0 && driver.part[2] >= 22000))
```

should be:

```c++
if (driver.part[0] > 31 || (driver.part[0] == 31 && driver.part[1] > 0) ||
      (driver.part[0] == 31 && driver.part[1] == 0 && driver.part[2] >= 22000))
```

CrystalP replied:
I see. Right, that was a typo. I replaced with a reusable function, which could be even more general, but that will be for later, if we need that driver version comparison somewhere else.

thexai commented on xbmc/rendering/dx/DeviceResources.cpp:
```diff
@@ -1484,3 +1484,20 @@ bool DX::DeviceResources::SetMultithreadProtected(bool enabled) const
+    return false;
+
+  return true;
+}
```
Missing empty line at EOF

CrystalP commented:
Issues resolved.

CrystalP committed:
[Windows] Add function to identify desktop AMD GCN gpu

thexai commented on xbmc/rendering/dx/DeviceResources.cpp:
```diff
@@ -1484,3 +1484,21 @@ bool DX::DeviceResources::SetMultithreadProtected(bool enabled) const
+
+  const DXGI_ADAPTER_DESC ad = GetAdapterDesc();
+
+  if (GetAdapterDesc().VendorId != PCIV_AMD)
```
```suggestion
  if (ad.VendorId != PCIV_AMD)
```
to avoid call  GetAdapterDesc() two times

thexai commented on xbmc/rendering/dx/DeviceResources.cpp:
```diff
@@ -1484,3 +1484,21 @@ bool DX::DeviceResources::SetMultithreadProtected(bool enabled) const
+
+  const VideoDriverInfo driver = CWIN32Util::GetVideoDriverInfo(ad.VendorId, ad.Description);
+
+  if (CWIN32Util::IsDriverVersionAtLeast(driver.version, "31.0.22000.0"))
```
Exceptionally driver info may be missing or corrupted in Windows Registry.

To avoid undefined behavior maybe check for valid in `VideoDriverInfo` struct before parse result:

```c++
  if (!driver.valid)
    return false;
```

CrystalP commented:
Both good ideas, done.
